lib/src/git.rs

```diff
@@ -685,6 +685,12 @@ pub fn reset_head(
 ) -> Result<(), git2::Error> {
     let first_parent_id = &wc_commit.parent_ids()[0];
     if first_parent_id != mut_repo.store().root_commit_id() {
+        if let Some(git_head) = mut_repo.view().git_head().as_normal() {
+            if git_head == first_parent_id {
+                return Ok(());
+            }
+        }
+
```

Comment:
Perhaps, we'll still need `git_repo.reset()` to reset the staging area. `jj co @-` doesn't move the HEAD, but it should have a clean working copy.

Reply:
Good point. Would be good to have a test for that case.

Reply:
From what I observed, `jj co @-` still has a clean working copy even without `reset`ting, but I have only done a couple of simple tests, is there any case when we actually need it?

Also, `reset` also takes in a `target` commit to which it [moves the `HEAD`](https://docs.rs/git2/latest/git2/struct.Repository.html#method.reset) (is `set_head_detached` even necessary in that case?), so I'm unsure how that could work. OTOH I was only able to fix the failing test by resetting.

Reply:
Maybe it only happens if you run `git add` or something like that and then `jj co @-`? I never used colocated repos. I actually didn't even remember if we would make the index match `@` or not (we don't) :)

Reply:
oh that's true, if I `git add` and get rid of the reset call things get pretty weird... but the question remains: what should we reset it to? if I simply reset to `first_parent` (i.e. `new_git_commit`) we still lose the branch we were pointing to before.

Reply:
Hmm. Maybe we need to call `repo.index().read_tree()` (https://docs.rs/git2/latest/git2/struct.Index.html#method.read_tree)

Reply:
> if I simply reset to `first_parent` (i.e. `new_git_commit`) we still lose the branch we were pointing to before.

Doesn't libgit2 move the branch target in that case?

If we need to do that manually, maybe we can extract the index handling to the common code path.
```rust
    if ... {
        git_repo.set_head_detached(new_git_commit_id)?;
    } else {
        git_repo.reference_symbolic("HEAD", UNBORN_ROOT_REF_NAME, true, "unset HEAD by jj")?;
    let mut index = git_repo.index()?;
    index.read_tree(...)?;
    index.write()?;
    git_repo.cleanup_state()?;
    mut_repo.set_git_head_target(new_target);
```    

> Maybe we need to call `repo.index().read_tree()` (https://docs.rs/git2/latest/git2/struct.Index.html#method.read_tree)

fwiw, the tree to reset might be the `rewrite::merge_commit_trees()` one if `@` is merged.

Reply:
I don't follow exactly how to get the correct `tree`s. Also, to check whether `@` is merged, should I simply look at the length for `wc_commit.parents()` (then pass that into `rewrite::merge_commit_trees`) or should I `match wc_commit.tree()?`? Also, on the `else` branch, is there a tree to read from?

Reply:
> to check whether `@` is merged, should I simply look at the length of `wc_commit.parents()` (then pass that into `rewrite::merge_commit_trees`)

You can just use the result of `rewrite::merge_commit_trees()`. It does nothing if `commits.len() == 1`. That said, I'm not pretty sure if teaching auto-merged tree to git is the right thing. I don't remember how git represents a merged (but not committed) working copy.

Anyway, using `git_repo.reset()` is the easiest fix. I think it would copy the behavior of `git reset` command and wouldn't rewrite symbolic HEAD unless necessary.

Reply:
You're right, I think I had tested with an invalid case so it really didn't work (`jj describe @-`, which I thought could also preserve HEAD, sorry for the confusion here).